philsh.1.txt
:man source: philsh 0.2
:man manual: Philsh's manual
philsh(1)
=========

NAME
----

philsh - Simple and lightweight shell

SYNOPSIS
--------

*philsh* [*-c* | *--execute* 'command'] [*-h* | *--help*] [*-v* | *--version*]

DESCRIPTION
-----------

Philsh is a simple shell.

OPTIONS
-------

*-c*, *--execute* 'command'::
 Execute a command, then exit.
*-h*, *--help*::
  Print help information to standard output, then exit.
*-v*, *--version*::
  Print philsh version to standard output, then exit.

BUILTIN COMMANDS
----------------

*cd*::
 Change the working directory.
*pwd*::
 Print current working directory.
*env*, *env* ['command']::
 Print environment variables or run 'command'.
*which*, *which* [*-a* | *--all*] 'program'::
 Print full path of 'program'
*jobs*::
 Print child process list who are runnig.
*alias*, *alias* [*-L*] ['alias_name[=string]']::
 Define or display aliases.
*unalias*, *unalias* [*-a*] ['alias_name']::
 Remove alias definition.
*uname*, *uname* [*-a* | *-s* | *-n* | *-r* | *-v* ]::
 Print system information.
*whoami*::
 Print effective user id.
*help*::
 Print informations about the philsh project (in french ;).
*exit*, *exit* ['return value']::
 Exit philsh, return 0 or value if it's set

CONFIGURATION
-------------

*philsh* is configured by '$HOME/.philshrc' for user configuration and by
'/etc/philsh/philshrc' for global configuration.
Theses configurations files must be simple philsh scripts. An example is located
in '/usr[/local]/share/philsh/philshrc.example'.

BUGS
----

Philsh isn't sable for now, so use it at your own risk. :)
Here is somes Known bugs.

*Builtin cd and pwd command*, 'src/int/cd.c' & 'src/int/pwd.c'::
 Theses functions don't support symbolics links.


AUTHORS
-------

Marc Lagrange <markocpc@gmail.com> for compilation system with cmake, git
repository.

Philippe Pepiot <philippe.pepiot@gmail.com> for the code.

WWW
---

Git repository at <http://tux-atome.fr/philsh.git/[]>

COPYING
-------

philsh is under BSD licence, see LICENCE file for more informations.

SEE ALSO
--------

Somes others shells : 'bash(1)', 'zsh(1)', 'sh(1)'...


src/philsh.c
/*
 * Copyright (C) 2008 Philippe Pepiot <philippe.pepiot@gmail.com>
 * philsh is under BSD licence, see LICENCE file for more informations.
 *
 */ 

/* on inclu les headers */
#include "headers.h"

/* {{{ variables globales */

/* Structure des builtin commandes,
 * char *name;
 * int (*p)(int, char**);
 */
const struct builtin builtin_command[] =
{
   {"pwd", internal_pwd, SAME_PROCESS},
   {"which", internal_which, NEW_PROCESS},
   {"uname", internal_uname, NEW_PROCESS},
   {"env", internal_env, NEW_PROCESS},
   {"cd", internal_cd, SAME_PROCESS},
   {"whoami", whoami, SAME_PROCESS},
   {"alias", alias, SAME_PROCESS},
   {"unalias", unalias, SAME_PROCESS},
   {"source", source, SAME_PROCESS},
   {"exit", exit_philsh, SAME_PROCESS},
   {"jobs", jobs, SAME_PROCESS},
   {"help", help, SAME_PROCESS},
   {NULL, NULL, NEW_PROCESS}
};
/* }}} */

/* {{{ init_env */
/* Cette fonction initialise l'environnement
 * Si $HOME et $PATH n'existent pas, il faut
 * les créer... En effet, si on lance le programme
 * comme ceci : env -i philsh, tous les getenv()
 * retournent NULL */
void init_env(void)
{
   /* L'uid sous lequel tourne philsh */
   uid_t uid = getuid();
   struct passwd *user;
   /* On recupère l'entrée correspondante
    * dans /etc/passwd */
   user = getpwuid(uid);
   /* On met tout ça dans l'environnement */
   setenv("HOME", user->pw_dir, 0);
   /* Si le PATH n'est pas fixé, il est impossible de le deviner
    * Dans ce cas on utilise un PATH classique UNIX */
   if (getenv("PATH") == NULL)
      setenv("PATH", "/usr/local/bin:/usr/bin:/bin:/opt/bin:/sbin", 0);
   return;
}
/* }}} */

void philsh(void)
{
   int fd_stdin, fd_stdout, ret = 0;
   char *p, *prompt, saisie[256];
   file_instruction *liste_instruction = NULL;
   for(;;)
   {
      prompt = set_prompt(ret);
      printf("%s", prompt);
      fflush(stdin);
      fgets(saisie, 256, stdin);
      fflush(stdin);
      fd_stdout = dup(1);
      fd_stdin = dup(0);
      p = strchr(saisie, '\n');
      *p = '\0';
      liste_instruction = creat_liste_instruction(saisie);
      ret = exec_file(liste_instruction);
#ifdef DEBUG
      afficher_liste_instruction(liste_instruction);
#endif
      close(1);
      dup(fd_stdout);
      close(0);
      dup(fd_stdin);
      free_file_instruction(liste_instruction);
   }
   return;
}

/* {{{ main() */
/* Ici commence l'aventure */
int main (int argc, char **argv)
{
   char *config_file;
   uid_t uid = getuid();
   struct passwd *user = getpwuid(uid);
   /* On teste les options avec lesquelles
    * philsh est appellé */
   options_philsh(argc, argv);
   /* On initialise l'environnement */
   init_env();
   /* Initialisation de la config */
   init_config("/etc/philsh/philshrc");
   config_file = malloc(sizeof(char) * (15+strlen(user->pw_dir)));
   sprintf(config_file, "%s/%s", user->pw_dir, ".philshrc");
   init_config(config_file);
   free(config_file);
   philsh();
   return 0;
}

/* }}} */

/* {{{ options_philsh */
int options_philsh(int argc, char **argv)
{
   file_instruction *cmd;
   int ret;
   const char *optstring = "hvc";
   static struct option philsh_option[] =
   {
      {"help", 0, NULL, 'h'},
      {"version", 0, NULL, 'v'},
      {"execute", 0, NULL, 'c'},
      {0, 0, 0, 0}
   };
   char opt;
   while (EOF != (opt = getopt_long(argc, argv, optstring, philsh_option, NULL)))
   {
      if (opt == 'v')
      {
	 printf("Philsh version : "PHILSH_VERSION".\n"
	       " Compilation settings :\n"
	       "  - Flags : "PHILSH_COMPILE_FLAGS"\n"
	       "  - Linked Libs : "PHILSH_LINKED_LIBS"\n"
	       "  - On "PHILSH_COMPILE_MACHINE" by "PHILSH_COMPILE_BY"\n"
	       " Main contact : "PHILSH_MAIL"\n");
	 exit(0);
      }
      else if(opt == 'h')
      {
	 afficher_aide();
	 exit(0);
      }
      else if(opt == 'c')
      {
	 if(argc < 3)
	    exit(0);
	 cmd = creat_liste_instruction(argv[2]);
	 ret = exec_file(cmd);
#ifdef DEBUG
	 afficher_liste_instruction(cmd);
#endif
	 free_file_instruction(cmd);
	 exit(ret);
      }
   }
   return 0;
}

/* }}} */

/* {{{ afficher_aide */

void afficher_aide(void)
{

   printf("PHILSH version "PHILSH_VERSION"\n\
	 philsh [OPTION]...\n\
	 -c, --execute commande (arguments)   Executer la commande puis quitter\n\
	 -h, --help                           afficher l'aide\n\
	 -v, --version                        afficher le nom et la version du logiciel\n\
	 man philsh pour voir la page de manuel\n\
	 \n\
	 <[HISTORIQUE]>\n\
	 Philsh est un shell minimaliste, écrit par Philippe Pepiot,\n\
	 étudiant en 3ème année de Mathématiques à Toulouse,\n\
	 qui à décidé d'apprendre le C sur internet et avec les pages de manuel.\n\
	 Philippe avait besoin de mettre en pratique ses maigres connaissances en C\n\
	 donc il s'est lancé dans le projet de faire un shell inutile, mais marrant.\n\
	 Le but premier est de coder, n'importe quoi pourvu que ce soit (marrant|instructif)\n\
	 Plus tard, il montre son code à rhaamo qui lui confectionne un systeme de compilation\n\
	 plus performant avec cmake. rhaamo lui apprend aussi les joie de git, et lui\n\
	 donne un depot git (git://tux-atome.fr/philsh.git)\n\
	 \n\
	 <[CONTACT]>\n\
	 Vous pouvez contacter les auteurs par mail :\n\
	 Philippe Pepiot alias philpep "PHILSH_MAIL"\n\
	 rhaamo <markocpc@gmail.com>\n\
	 <"PHILSH_HTTP">\n\
	 \n\n\n");
   return;
}

/* }}} */
/* vim:fdm=marker:
*/
